Bitcoin assets (#1188)
dfinity/examples: rust/basic_bitcoin/README.md

This example demonstrates how to deploy a smart contract on the Internet Computer that can receive and send Bitcoin, including support for legacy (P2PKH), SegWit (P2WPKH), and Taproot (P2TR) address types.

The repository also includes examples of how to work with Bitcoin assets such as Ordinals, Runes, and BRC-20 tokens.

## Table of Contents

- [Architecture](#architecture)
- [Prerequisites](#prerequisites)
- [Building and deploying the smart contract](#building-and-deploying-the-smart-contract)
- [Generating Bitcoin addresses](#generating-bitcoin-addresses)
- [Receiving Bitcoin](#receiving-bitcoin)
- [Checking balance](#checking-balance)
- [Sending Bitcoin](#sending-bitcoin)
- [Retrieving block headers](#retrieving-block-headers)
- [Bitcoin Assets](#bitcoin-assets)
- [Inscribe an Ordinal](#inscribe-an-ordinal)
- [Etch a Rune](#etch-a-rune)
- [Deploy a BRC-20 Token](#deploy-a-brc-20-token)
- [Notes on implementation](#notes-on-implementation)

> - [Security considerations and best practices](#security-considerations-and-best-practices)

Reviewer: The capitalization is inconsistent. Sometimes it's title case, sometimes it's not (for example, "Assets" is capitalized but "balance" is not). Could this be made consistent?

Author: Fixed

Author: Yeah @jessiemongeon1, honestly don't remember what we said here. We reference both bitcoin the network and bitcoin the currency. 

For Rune and Ordinal, I'd say they should be capitalized as the "brand names" of those.. things.

Token and asset should have no special treatment as you say, I just missed them when correcting. Will fix.

Reviewer 2: We always follow our style guide: https://github.com/dfinity/portal/blob/master/style-guide.md#language

- Bitcoin: The network, software, architecture, etc.
- bitcoin: BTC
- Rune: Capitalized
- Ordinal: Capitalized.

- `dfx deploy` tells the command line interface to `deploy` the smart contract.
- `--argument '(variant { regtest })'` passes the argument `regtest` to initialize the smart contract, telling it to connect to the local Bitcoin regtest network.

> Your smart contract is live and ready to use! You can interact with it using either the command line or the Candid UI, which is the link you see in the terminal.

Reviewer 2: ```suggestion
Your smart contract is live and ready to use! You can interact with it using either the command line or the Candid UI (the link you see in the terminal).
```

# or get_p2wpkh_address, get_p2tr_key_path_only_address, get_p2tr_script_path_enabled_address
```

## Receiving bitcoin

> Use the `bitcoin-cli` to mine a Bitcoin block and send the block reward in the form of local testnet bitcoin (tBTC) to one of the smart contract addresses.

Reviewer 2: tBTC appears to be its own token: https://coinmarketcap.com/currencies/tbtc/

Could this be confusing?

This calls `bitcoin_get_block_headers`, useful for validating blockchains or light client logic.

> ## Inscribe an ordinal

Reviewer 2: ```suggestion
## Inscribe an Ordinal
```

> Make sure bitcoind is configured to accept non standard transactions.

Reviewer 2: ```suggestion
:::caution

Ensure `bitcoind` is configured to accept non-standard transactions.

:::
```

Author: I will add secttion using GitHuib markdown format instead. `> [!WARNING]`

> The `acceptnonstdtxn` setting needs to be included in your `bitcoin.conf` file. Example:

Reviewer 2: ```suggestion
The `acceptnonstdtxn` setting needs to be included in your `bitcoin.conf` file, for example:
```

```
regtest=1
acceptnonstdtxn=1
txindex=1
rpcuser=<username>
rpcpassword=<password>
```

```

> ord --config-dir . server

Reviewer 2: These commands need an introduction

dfx canister call basic_bitcoin get_p2tr_key_path_only_address '()'

bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 100 <p2tr_key_path_only_address>

dfx canister call basic_bitcoin inscribe_ordinal '("ORDINAL")'

bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 1 <p2tr_key_path_only_address>

```

## Etch a Rune

> 

Reviewer 2: These commands need an introduction

Author: yes, haven't worked on those yet.

dfx canister call basic_bitcoin get_block_headers '(0: nat32, 11: nat32)'
```

This calls `bitcoin_get_block_headers`, which is useful for blockchain validation or light client logic.

## Bitcoin assets

Bitcoin's scripting capabilities enable various digital assets beyond simple transfers. This example demonstrates how to create and interact with three major Bitcoin asset protocols from an ICP smart contract:

- **Ordinals**: Inscribe arbitrary data onto individual satoshis

> - **Runes**: Create fungible tokens using OP_RETURN outputs

Reviewer 2: ```suggestion
- **Runes**: Create fungible tokens using `OP_RETURN` outputs
```

### Step-by-step process:

1. **Start the ord server** to index transactions:
   ```bash
   ord --config-dir . server
   ```

2. **Get a Taproot address** for funding the inscription:
   ```bash
   dfx canister call basic_bitcoin get_p2tr_key_path_only_address '()'
   ```

3. **Fund the address** with sufficient bitcoin (100 blocks ensures spendability):
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 100 <p2tr_key_path_only_address>
   ```

4. **Create the inscription** with your desired text:
   ```bash
   dfx canister call basic_bitcoin inscribe_ordinal '("Hello Bitcoin")'
   ```

5. **Mine a block** to confirm the transactions:
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 1 <p2tr_key_path_only_address>
   ```

The function returns the reveal transaction ID. Your inscription is now permanently stored on Bitcoin and can be viewed using ord or other Ordinals explorers. The default address of the local `ord` server is `http://127.0.0.1:80/`.

## Etch a Rune

> [Runes](https://docs.ordinals.com/runes.html) is a fungible token protocol that embeds token metadata directly into Bitcoin transactions using OP_RETURN outputs. Unlike Ordinals, Runes are created in a single transaction and support standard fungible token operations.

Reviewer 2: ```suggestion
[Runes](https://docs.ordinals.com/runes.html) is a fungible token protocol that embeds token metadata directly into Bitcoin transactions using `OP_RETURN` outputs. Unlike Ordinals, Runes are created in a single transaction and support standard fungible token operations.
```

### Step-by-step process:

> 1. **Start the ord server** to track rune balances:

Reviewer 2: ```suggestion
1. **Start the ord server** to track Rune balances:
```

   ```bash
   ord --config-dir . server
   ```

> 2. **Get a Taproot address** for the rune etching:

Reviewer 2: ```suggestion
2. **Get a Taproot address** for the Rune etching:
```

   ```bash
   dfx canister call basic_bitcoin get_p2tr_key_path_only_address '()'
   ```

3. **Fund the address** with bitcoin to pay for the etching:
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 100 <p2tr_key_path_only_address>
   ```

> 4. **Etch the rune** with an uppercase name (maximum 28 characters):

Reviewer 2: ```suggestion
4. **Etch the Rune** with an uppercase name (maximum 28 characters):
```

   ```bash
   dfx canister call basic_bitcoin etch_rune '("ICPRUNE")'
   ```

5. **Mine a block** to confirm the etching:
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 1 <p2tr_key_path_only_address>
   ```

> 6. **Decode the runestone** to verify the etching:

Reviewer 2: ```suggestion
6. **Decode the Runestone** to verify the etching:
```

   ```bash
   ord --config-dir . decode --txid <transaction_id>
   ```

> The rune is now etched with 1,000,000 tokens minted to your address. The tokens can be transferred using standard Bitcoin transactions with runestone data.

Reviewer 2: ```suggestion
The Rune is now etched with 1_000_000 tokens minted to your address. The tokens can be transferred using standard Bitcoin transactions with Runestone data.
```

## Deploy a BRC-20 token

[BRC-20](https://domo-2.gitbook.io/brc-20-experiment/) is a token standard built on top of Ordinals that uses structured JSON payloads to create fungible tokens. BRC-20 tokens follow the same inscription process as Ordinals but with standardized JSON formats.

### Step-by-step process:

1. **Start the ord server** to index BRC-20 inscriptions:
   ```bash
   ord --config-dir . server
   ```

2. **Get a Taproot address** for the token deployment:
   ```bash
   dfx canister call basic_bitcoin get_p2tr_key_path_only_address '()'
   ```

3. **Fund the address** with bitcoin:
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 100 <p2tr_key_path_only_address>
   ```

4. **Deploy the BRC-20 token** with a 4-character ticker:
   ```bash
   dfx canister call basic_bitcoin inscribe_brc20 '("DEMO")'
   ```

5. **Mine a block** to confirm the deployment:
   ```bash
   bitcoin-cli -conf=$(pwd)/bitcoin.conf generatetoaddress 1 <p2tr_key_path_only_address>
   ```

This creates a BRC-20 token with:
- Ticker: "DEMO"

> - Max supply: 21,000,000 tokens
> - Mint limit: 1,000 tokens per mint

Reviewer 2: ```suggestion
- Max supply: 21_000_000 tokens
- Mint limit: 1_000 tokens per mint
```
